### PR TITLE
[ADD] partner_aging

### PR DESCRIPTION
**Pull request by max3903**
This module creates new customer / supplier aging as of a given date.

The default Aged Partner balance report is related to specific date and a static PDF that is based on the difference between credits and debits, not based on documents such as Invoices and Payments. This module provides an interactive view with details of the invoice. This module does not consider unapplied credits.

# Usage

To use this module, you need to:

* Go to Accounting
* Click on Sales > Customer Aging or Purchases > Supplier Aging
* Change the date if necessary

https://github.com/OCA/l10n-usa/issues/5

### CODE REVIEW

File: partner_aging/wizard/customer.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                and aml.amount_residual != 0
+                GROUP BY aml.partner_id, aml.id, ai.number, days_due, ai.user_id, ai.id
+              """ % (age_date, age_date, age_date, age_date, age_date,  # noqa
+                     age_date, age_date, age_date, age_date, age_date,
+                     age_date, age_date, age_date, age_date)
+
+        tools.drop_view_if_exists(self.env.cr, '%s' %
+                                  (customer_aging._name.replace('.', '_')))
+        self.env.cr.execute("""CREATE OR REPLACE VIEW %s AS (%s)""" %
+                            (customer_aging._name.replace('.', '_'), query))
```
Comment (vprime) [inline partner_aging/wizard/customer.py]: This above query and block of code gets reused a few times, can we move this into it's own method?

**Review by vprime — CHANGES_REQUESTED** [review]
The aging query is complex and gets duplicated a few times, I would like to see that moved into it's own method.

File: partner_aging/wizard/customer.py (hunk 1 above)
Comment (smangukiya) [reply partner_aging/wizard/customer.py]: @vprime : Please review it again, I have make changes as per your instruction.

cc: @max3903

File: partner_aging/README.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+=============
+Partner Aging
+=============
+
+This module creates new customer / supplier aging as of a given date.
+
+The default Aged Partner balance report is related to specific date and a static PDF that is based on the difference between credits and debits, not based on documents such as Invoices and Payments. This module provides an interactive view with details of the invoice. This module does not consider unapplied credits.
+
+Installation
```
Comment (lasley) [inline partner_aging/README.rst]: Can remove empty sections

File: partner_aging/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+# Copyright 2017 Ursa Information Systems <http://www.ursainfosystems.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'Interactive Partner Aging at any date',
+    'summary': 'Aging as a view - invoices and credits',
+    'version': '10.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': 'Ursa Information Systems, Odoo Community Association (OCA)',
+    'category': 'Accounting & Finance',
+    'maintainer': 'Ursa Information Systems',
```
Comment (lasley) [inline partner_aging/__manifest__.py]: Do we use this key?

File: partner_aging/wizard/__init__.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Ursa Information Systems <http://www.ursainfosystems.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from . import (
```
Comment (lasley) [inline partner_aging/wizard/__init__.py]: use explicit imports:
```
from . import customer
from . import supplier
```

File: partner_aging/wizard/customer.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Ursa Information Systems <http://www.ursainfosystems.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import tools
+from odoo import api, fields, models, _
+
+
+class PartnerAgingDate(models.TransientModel):
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: [Split models by file](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#file-naming)

File: partner_aging/wizard/customer.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    _name = "partner.aging.date"
+
+    age_date = fields.Datetime("Aging Date",
+                               required=True,
+                               default=lambda self: fields.Datetime.now())
+
+    @api.multi
+    def open_partner_aging(self):
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: Please add a test for this

File: partner_aging/wizard/customer.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+# Copyright 2017 Ursa Information Systems <http://www.ursainfosystems.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import tools
+from odoo import api, fields, models, _
+
+
+class PartnerAgingDate(models.TransientModel):
+    _name = "partner.aging.date"
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: IMO this should be `res.partner.aging.date` to match with `res.partner` schema

File: partner_aging/wizard/customer.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            'view_mode': 'tree,form',
+            'res_model': 'partner.aging.customer.ad',
+            'type': 'ir.actions.act_window',
+            'domain': [('total', '!=', 0)],
+            'context': ctx,
+        }
+
+
+class AccountAgingCustomerAD(models.Model):
+    _name = 'partner.aging.customer.ad'
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: This name does not match the class name. Also, I'm kind of unclear on what this model does by the naming. I don't have a suggestion though, but would be happy to provide one if you elaborate on what you were going for?

File: partner_aging/wizard/customer.py (hunk 5 above)
Comment (lasley) [reply partner_aging/wizard/customer.py]: Also this is a wizard, so it should be in the `wizards` folder

File: partner_aging/wizard/customer.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    _auto = False
+
+    @api.multi
+    def execute_aging_query(self, age_date=False):
+        if not age_date:
+            age_date = fields.Date.context_today(self)
+
+        query = """
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: This is a lot of SQL that I feel can be simplified with some string templating in Python first. That's not a blocker, but the lack of testing on all this logic definitely is 👎

File: partner_aging/wizard/customer.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                     age_date, age_date, age_date, age_date, age_date,
+                     age_date, age_date, age_date, age_date)
+
+        tools.drop_view_if_exists(self.env.cr, '%s' %
+                                  (self._name.replace('.', '_')))
+        self.env.cr.execute("""CREATE OR REPLACE VIEW %s AS (%s)""" %
+                            (self._name.replace('.', '_'), query))
+
+    @api.multi
+    def docopen(self):
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: This should probably be named `open_document` to match the `open_partner_aging` schema.

File: partner_aging/wizard/customer.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        """
+        @description  Open document (invoice or payment) related to the
+                      unapplied payment or outstanding balance on this line
+        """
+
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: `self.ensure_one()`

File: partner_aging/wizard/customer.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        models = self.env['ir.model.data']
+        # Get this line's invoice id
+        inv_id = self.invoice_id.id
+
+        # If this is an unapplied payment (all unapplied payments hard-coded to
+        #  -999), get the referenced voucher
+        if inv_id == -999:
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: Where does this `-999` come from? It should be a constant.

File: partner_aging/wizard/customer.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+            payment = self.env['account.voucher']
+            # Get referenced customer payment
+            # (invoice_ref field is actually a payment for these)
+            voucher_id = payment.search([('number', '=', self.invoice_ref)])[0]
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: This will throw an `IndexError` if there's no voucher

File: partner_aging/wizard/customer.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            view = models.xmlid_to_object('account_voucher.view_voucher_form')
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: Why not just `self.env.ref`?

File: partner_aging/wizard/customer.py (hunk 10 above)
Comment (lasley) [reply partner_aging/wizard/customer.py]: Also this should be tested

File: partner_aging/wizard/customer.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            # Set values for form
+            view_id = view and view.id or False
+            name = 'Customer Payments'
+            res_model = 'account.voucher'
+            ctx = "{}"
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: A string? IMO this should be a real dict & Python will take care of string translation if/when necessary.

File: partner_aging/wizard/customer.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+            'view_id': [view_id],
+            'res_model': res_model,
+            'context': ctx,
+            'type': 'ir.actions.act_window',
+            'nodestroy': True,
+            'target': 'current',
+            'res_id': doc_id,
+        }
+
+    partner_id = fields.Many2one('res.partner', u'Partner', readonly=True)
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: [Field declarations should be at top.](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#models)

Also, please remove all the forced unicodes

File: partner_aging/wizard/customer.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+    avg_days_overdue = fields.Integer(u'Avg Days Overdue', readonly=True)
+    date = fields.Date(u'Date', readonly=True)
+    date_due = fields.Date(u'Due Date', readonly=True)
+    inv_date_due = fields.Date(u'Invoice Date', readonly=True)
+    total = fields.Float(u'Total', readonly=True)
+    not_due = fields.Float(u'Not Due Yet', readonly=True)
+    days_due_01to30 = fields.Float(u'1/30', readonly=True)
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: These fields are not PEP-8 compliant. Please do not deviate from snake casing

File: partner_aging/wizard/customer.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+    days_due_31to60 = fields.Float(u'31/60', readonly=True)
+    days_due_61to90 = fields.Float(u'61/90', readonly=True)
+    days_due_91to120 = fields.Float(u'91/120', readonly=True)
+    days_due_121togr = fields.Float(u'+121', readonly=True)
+    max_days_overdue = fields.Integer(u'Days Outstanding', readonly=True)
+    invoice_ref = fields.Char('Our Invoice', size=25, readonly=True)
+    invoice_id = fields.Many2one('account.invoice', 'Invoice', readonly=True)
+    salesman = fields.Many2one('res.users', u'Sales Rep', readonly=True)
+
+    _order = 'partner_id'
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: This should be at top

File: partner_aging/wizard/customer.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+    @api.model_cr
+    def init(self):
+        self.execute_aging_query()
```
Comment (lasley) [inline partner_aging/wizard/customer.py]: What about the super init?

File: partner_aging/wizard/supplier.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Ursa Information Systems <http://www.ursainfosystems.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import tools
+from odoo import api, fields, models, _
+
+
+class SupplierAgingDate(models.TransientModel):
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: Split models by file

File: partner_aging/wizard/supplier.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+            'view_mode': 'tree,form',
+            'res_model': 'partner.aging.supplier.ad',
+            'type': 'ir.actions.act_window',
+            'domain': [('total', '!=', 0)],
+            'context': ctx,
+        }
+
+
+class PartnerAgingSupplierAD(models.Model):
+    _name = 'partner.aging.supplier.ad'
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: Same question regarding the naming. What is AD? 

General note - abbreviations are typically bad and obscure the code. Unless something is insanely long, it's best to just use the full term everywhere.

File: partner_aging/wizard/supplier.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+    _auto = False
+
+    @api.multi
+    def execute_aging_query(self, age_date=False):
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: All this needs to be tested

File: partner_aging/wizard/supplier.py [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+                     age_date, age_date, age_date, age_date, age_date,
+                     age_date, age_date, age_date, age_date)
+
+        tools.drop_view_if_exists(self.env.cr, '%s' %
+                                  (self._name.replace('.', '_')))
+        self.env.cr.execute("""CREATE OR REPLACE VIEW %s AS (%s)""" %
+                            (self._name.replace('.', '_'), query))
+
+    @api.multi
+    def invopen(self):
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: `open_invoice`

File: partner_aging/wizard/supplier.py [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+            'view_id': [view_id],
+            'res_model': 'account.invoice',
+            'context': "{'type':'in_invoice'}",
+            'type': 'ir.actions.act_window',
+            'nodestroy': True,
+            'target': 'current',
+            'res_id': self.invoice_id.id
+        }
+
+    partner_id = fields.Many2one('res.partner', u'Partner', readonly=True)
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: Move to top & remove unicode

File: partner_aging/wizard/supplier.py [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+    not_due = fields.Float(u'Not Due Yet', readonly=True)
+    days_due_01to30 = fields.Float(u'1/30', readonly=True)
+    days_due_31to60 = fields.Float(u'31/60', readonly=True)
+    days_due_61to90 = fields.Float(u'61/90', readonly=True)
+    days_due_91to120 = fields.Float(u'91/120', readonly=True)
+    days_due_121togr = fields.Float(u'+121', readonly=True)
+    invoice_ref = fields.Char('Their Invoice', size=25, readonly=True)
+    invoice_id = fields.Many2one('account.invoice', 'Invoice', readonly=True)
+
+    _order = 'partner_id'
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: Move to top

File: partner_aging/wizard/supplier.py [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+
+    @api.model_cr
+    def init(self):
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: What about super init?

File: partner_aging/wizard/supplier.py [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+            'view_type': 'form',
+            'view_mode': 'tree,form',
+            'res_model': 'partner.aging.supplier.ad',
+            'type': 'ir.actions.act_window',
+            'domain': [('total', '!=', 0)],
+            'context': ctx,
+        }
+
+
+class PartnerAgingSupplierAD(models.Model):
```
Comment (lasley) [inline partner_aging/wizard/supplier.py]: I feel like this duplicates a significant amount of code with the customer side of things. I would recommend using an abstract model to unify this

**Review by lasley — DISMISSED** [review]
Thanks @max3903 - comments inline.

From a high level, I am blocking due to lack of testing. There is a significant amount of logic here, some of which is direct SQL, but there are zero automated tests 👎

**Comment by max3903** [comment]
:+1:

**Review by bodedra — APPROVED** [review]
I improved some existing code.

LGTM. :+1: 

Minor correction:

@Sudarshan1607 can you update Copyrights ? Some files I see 2012, 2017 and 2018.

File: partner_aging/README.rst [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* Bhavesh Odedra <bodedra@ursainfosystems.com>
+* Balaji Kannan <bkannan@ursainfosystems.com>
+* Sandeep Mangukiya <smangukiya@ursainfosystems.com>
+
```
Comment (bodedra) [inline partner_aging/README.rst]: @Sudarshan1607 Add yourself to the list

Comment (smangukiya) [reply partner_aging/README.rst]: Done! along with copyright info

**Review by smangukiya — APPROVED** [review]
Seems ok to go!

**Pull request merged**